### PR TITLE
Address infinite loop with floating point inf to Decimal

### PR DESCRIPTION
**Pull request by mita4829**
Swift currently does not support infinites for the `Decimal` class. Because of this, any code that tries to convert a floating point value of `Inf` or `-Inf` to `Decimal` will cause the compiler to go into an infinite loop due to the existing implementation of the `Decimal` class. 

**Repro**
```
let inf= Double.infinity
let decimal = Decimal(floatLiteral: inf) // <-- Infinite loop
```

As an alternative till infinity is properly defined in the `Decimal` class, one is to catch these infinity conversions and turn them into `Decimal.nan`

Resolves [SR-3126](https://bugs.swift.org/browse/SR-3126).

### CODE REVIEW

**Comment by mita4829** [comment]
@swift-ci Please smoke test

**Comment by phausler** [comment]
I would almost expect it to precondition failure on infinities since they are not representable in Decimal.

**Comment by mita4829** [comment]
A precondition failure doesn't exist as of today when infinities are passed to the Decimal class. Setting infinities to `Decimal.nan` seems to be the cleanest mitigation for the loop during compile time and runtime until infinities are properly defined.

**Comment by xwu** [comment]
I agree it should be a precondition failure. Behavior does not have to match the Decimal class.

**Review by phausler — APPROVED** [review]
This seems more accurate to me and informs the developer the truth before blowing up in a NaN calculation.

File: stdlib/public/Darwin/Foundation/Decimal.swift [hunk 1]
```diff
@@ -314,9 +314,10 @@ extension Decimal {
     }
     
     public init(_ value: Double) {
-        if value.isNaN || value == Double.infinity || value == -Double.infinity {
-            // Note: Decimal does not have infinities yet.
+        if value.isNaN {
             self = Decimal.nan
+        } else if value == Double.infinity || value == -Double.infinity {
```
Comment (xwu) [inline stdlib/public/Darwin/Foundation/Decimal.swift]: Can I suggest `else if value.isInfinite`?

**Comment by CodaFi** [comment]
Resolved by the merge of 3104ccb.  Thanks!

**Pull request closed without merging**